app/javascript/components/common/Editor/TestRuns.tsx

```diff
@@ -0,0 +1,57 @@
+import React, { useState, useEffect } from 'react'
+import { Submission, SubmissionTestsStatus } from '../Editor'
+import consumer from '../../../utils/action-cable-consumer'
+
+function Content({ submission }: { submission: Submission }) {
+  switch (submission.testsStatus) {
+    case SubmissionTestsStatus.PASS:
+    case SubmissionTestsStatus.FAIL:
+      return (
+        <>
+          {submission.testRuns.map((run) => (
+            <p key={run.name}>
+              name: {run.name}, status: {run.status}, output: {run.output}
+            </p>
+          ))}
+        </>
+      )
+    case SubmissionTestsStatus.ERROR:
+    case SubmissionTestsStatus.OPS_ERROR:
+      return <p>{submission.message}</p>
+    default:
+      return <></>
+  }
+}
+
+export function TestRuns(props: { submission: Submission }) {
+  const [submission, setSubmission] = useState(props.submission)
+
+  useEffect(() => {
+    const subscription = consumer.subscriptions.create(
+      { channel: 'Test::Submissions::TestRunsChannel', id: submission.id },
+      {
+        received: (json: any) => {
```

Comment:
@SleeplessByte Is there a good way to do the ff:

1. Camelize the keys from the JSON response.
2. Convert the JSON response into one of our types.

I'm doing it manually here now, but I'm unsure if there's a better way.

Reply:
Yeah.

1. Run [`camel-case`](https://www.github.com/blakeembrey/change-case/tree/master/packages%2Fcamel-case%2FREADME.md) recursively.
2. We can assert then type cast:

```typescript
function typecheck<T>(json: any, key: string): T | never {
  // For simplicity, let's say that we trust the
  // JSON if it's a non-empty object. In reality
  // we'd want to use media types to ensure
  // this. I recommend the above regardless
  // so structure all JSON results with at least
  // a single top level key, e.g. exercise,
  // concepts, messages, user:
  //
  // {
  //    "exercise": ...
  // }
  
  if (json === null || typeof json !== 'object') {
    throw new Error(`Expexted non-null object, actual ${json === null ? null : typeof json}`)
  }

  if (!json.hasOwnProperty(key)) {
    const keys = Object.keys(json)
    throw new Error(`Expected top-level key ${key}, actual keys: ${keys}`)
  }
  
  // You can either decide to type the value of 
  // the top level key, and return the value,
  // or if you want more than one top-level key
  // type the entire thing.
  return json[key] as T
}
```

Typed this on my phone, apologies for errors